### PR TITLE
Deps+Everywhere: Upgrade `discord.js` to `v14`

### PR DESCRIPTION
**Pull request by filiphsps**
Additionally, every command now
  - uses the `SlashCommandBuilder` instead of directly providing JSON.
  - have their handlers properly typed.

Some parts of this commit are based on @bgianfo's work in #647.

### CODE REVIEW

**Comment by bgianfo** [comment]
You have no idea how happy I am to see this PR, thank you! :blue_heart: :green_heart: :purple_heart: :yellow_heart:

File: src/commandHandler.ts [hunk 1]
```diff
@@ -85,24 +87,28 @@ export default class CommandHandler {
...
     /** Executes user commands contained in a message if appropriate. */
-    async handleBaseCommandInteraction(interaction: BaseCommandInteraction): Promise<void> {
+    async handleCommandInteraction(interaction: Interaction): Promise<void> {
         if (!this.production) {
-            const msg = `Buggie bot received '${JSON.stringify(interaction, (_, v) =>
-                typeof v === "bigint" ? `${v.toString()}n` : v
+            const msg = `Buggie bot received ${JSON.stringify(
+                interaction,
+                (_, v) => (typeof v === "bigint" ? `${v.toString()}n` : v),
+                4
```
Comment (bgianfo) [inline src/commandHandler.ts:74]: Can we use named arguments for this to make it clear this is for the number of spaces to use as indention?

File: src/commands/userCommand.ts [hunk 2]
```diff
@@ -101,30 +98,31 @@ export class UserCommand extends Command {
...
                                 };
                             })
                         )
                 )
             );
-        }
-        if (rows.length === 0) return;
-        const embed = new MessageEmbed()
+
+        if (rows.length === 0) throw new Error("");
```
Comment (bgianfo) [inline src/commands/userCommand.ts:109]: Real error message?

**Review by bgianfo — CHANGES_REQUESTED** [review]
Just some questions and small nit picks.

File: src/util/embedFromIssueOrPull.ts [hunk 3]
```diff
@@ -119,36 +137,58 @@ export function embedFromPull(
...
-        .addField("Comments", pull.comments.toString(), true)
-        // @ts-expect-error Intl.ListFormat is not yet part of the typescript library definitions,
-        //                  see https://github.com/microsoft/TypeScript/issues/46907
-        .addField("State", new Intl.ListFormat().format(state), true);
+        .setDescription(description || null)
+        .addFields(
+            { name: "Type", value: "Pull Request", inline: true },
+            {
+                name: "Created",
+                value: `<t:${new Date(pull.created_at).valueOf() / 1000}:R>`,
```
Comment (bgianfo) [inline src/util/embedFromIssueOrPull.ts:149]: Is there a better way to do this that discord provides? We don't need to do it in this PR, but it seems like we should probably factor this out into a utility if `discord.js` doesn't provide one?

File: src/commandHandler.ts [hunk 4]
```diff
@@ -34,22 +25,10 @@ export default class CommandHandler {
...
-            EmojiCommand,
-            QuoteCommand,
-            UserCommand,
-        ];
-
         const availableCommands = new Array<string>();
 
         this.commands = new Map(
-            commandClasses.map(commandClass => {
+            Object.values(Commands).map(commandClass => {
```
[the remark this answers is not in the document]
Comment (filiphsps) [reply src/commandHandler.ts:31]: It iterates over everything that’s (named and) “exported” from the commands file without any safe guards, hypothetically you could export the Command base class in the same way and break it, but that doesn’t fit our current structure would be catched with a test.

File: src/util/embedFromIssueOrPull.ts (hunk 3 above)
Comment (filiphsps) [reply src/util/embedFromIssueOrPull.ts:149]: This isn’t something that was done as a part of this upgrade, it existed beforehand. With that said; yes there might be a way to do this “properly”, I’ll take a look as a while fixing the error messages :)

File: src/commandHandler.ts (hunk 1 above)
Comment (filiphsps) [reply src/commandHandler.ts:74]: You’re right, I completely forgot the best thing to ever happen to typescript; named parameters.

Comment (filiphsps) [reply src/commandHandler.ts:74]: Is that a thing in typescript now? If that’s the case then have completely missed that, awesome!

File: src/util/embedFromIssueOrPull.ts (hunk 3 above)
Comment (filiphsps) [reply src/util/embedFromIssueOrPull.ts]: Turns out that yes; there is a utility function for it!

File: src/commandHandler.ts (hunk 1 above)
Comment (filiphsps) [reply src/commandHandler.ts:74]: > Can we use named arguments for this to make it clear this is for the number of spaces to use as indention?

I looked into "named arguments/parameters" isn't really a thing in js/ts, the usual way to handle this is to take an object as input instead; however, this is a built-in function that doesn't support that.

File: src/commandHandler.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (filiphsps) [reply src/commandHandler.ts:31]: > How does this avoid the abstract `Command` base class that everything derives from? I honestly don't know that much javascript/typescript, so mostly asking out of curiosity.

Basically, what we get from `Commands` is the following:
```javascript
{
  CommitStatsCommand: [Getter],
  EmojiCommand: [Getter],
  GithubCommand: [Getter],
  ManCommand: [Getter],
  PlanCommand: [Getter],
  QuickLinksCommand: [Getter],
  QuoteCommand: [Getter],
  Test262Command: [Getter],
  UserCommand: [Getter]
}
```

Unless we start exporting `Command` as a non-default it won't be a part of the `Object.values(Commands)` array/list.

File: src/commandHandler.ts (hunk 1 above)
Comment (bgianfo) [reply src/commandHandler.ts:74]: Sure, I guess I was thinking about explicit object member names. Sorry about that.

File: src/util/embedFromIssueOrPull.ts (hunk 3 above)
Comment (bgianfo) [reply src/util/embedFromIssueOrPull.ts]: cool! thanks

**Review by bgianfo — APPROVED** [review]

**Pull request merged**